cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -3162,9 +3187,13 @@ class AzureFileSystem::Impl {
       return Status::OK();
     }
     std::string sas_token;
-    {
+    if (options_.UsesSasCredential()) {
+      // When authenticated with SAS token GetUrl() automatically provides an
+      // authenticated URL with the SAS token appended.
+      sas_token = "";
```

Comment:
How about changing `GenerateSASToken()` instead? If we do, we don't need to define `UsersSasCredential()`.

```diff
diff --git a/cpp/src/arrow/filesystem/azurefs.cc b/cpp/src/arrow/filesystem/azurefs.cc
index 78f4ad1edd..c01ca41aed 100644
--- a/cpp/src/arrow/filesystem/azurefs.cc
+++ b/cpp/src/arrow/filesystem/azurefs.cc
@@ -410,6 +410,10 @@ AzureOptions::MakeDataLakeServiceClient() const {
 
 Result<std::string> AzureOptions::GenerateSASToken(
     Storage::Sas::BlobSasBuilder* builder, Blobs::BlobServiceClient* client) const {
+  if (credential_kind_ == CredentialKind::kSasToken) {
+    // TODO: Write why we can return "" here
+    return "";
+  }
   using SasProtocol = Storage::Sas::SasProtocol;
   builder->Protocol =
       blob_storage_scheme == "http" ? SasProtocol::HttpsAndHttp : SasProtocol::HttpsOnly;
```